### DISCUSSION TITLE
Please help, I modified the crazyflie firmware to make a new version of PID

### DISCUSSION THREAD

**Discussion by loraine92** [bitcraze/discussions#1242] [Q&A]
Please help me, I modified the crazyflie firmware to make a new version of PID, I have a question if I should only change the values of the position and velocity PID constants in the platform_defaults_cf2 file or should I also modify those values in platform_defaults_bolt, platform_defaults_flapper , platform_defaults_tag?

The PID I want to make is the one shown in the image, can you give me some advice on how to do it, please.
![PID PARALELO MATLAB](https://github.com/bitcraze/discussions/assets/80229306/521c8906-66dd-4c96-a53b-7ce547ec2239)

**Comment by knmcguire** [comment]
If you just need to tune the gains I don't think you need to implement a whole new PID controller. Just mind that on the crazyflie a cascaded PID controller is implemented: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/#cascaded-pid-controller.

So if you want to simplify that, namely position to attitude ref, then you'll need to do the following: 
* Change the gains of the velocity PID and the altitude rate PID such there is only forwarding: So that means the I and D gain to 0 and P gain to 1. You can do that through the[ parameter change of these](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/parameters_tab/) in real time.
   * [velCtlPiD]( https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#velctlpid)
   * [pid_rate](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#pid_rate)
* Change the gains of your position PID (where your position ref comes from) and your attitude PID
   * [posCtlPid](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#posctlpid)
   * [pid_attitude](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#pid_attitude)
You can set these parameters to be persistent through the same parameters tab, such that these values retain after restart. 

Just mind that I don't expect these changes will make your crazyflie fly. A quadcopter is a highly nonlinear unstable system, and if you simulink model is describing a linear decoupled system  like it does now, these gains will not translate well in the real world. There is a reason why the PID controller is as heavily cascaded as it is. 

Two things that might also be handy to know:
* The cfclient has a [tuning tab](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/tuning_tab/) that might be useful here
* There is a more complex crazyflie simulation in Webots that you can use instead. It is able to take matlab code so perhaps you can use this to link to your simulink model. Just go through this [getting started tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-simulation/)

**Reply by loraine92** [reply]
Thanks for answering, I want to ask you another question, in the script position_controller_pid.c it has the functions that perform position and velocity control, I want to see the data of setpoint_velocity.x and setpoint_velocity.y to compare that data with my matlab simulation of the PID in cascade, could you explain to me how I could send the data of those variables to the microSD memory. Thank you so much.

**Reply by knmcguire** [reply]
The instructions of using the sd card deck can be found here : https://www.bitcraze.io/products/micro-sd-card-deck/

also you can use the plotting functionality of the cfclient: https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/plotter_tab/

If you have any other questions, we usually prefer one discussion thread per question. So if you could select 'mark as answer' as well to close this?
